modality-probe-cli/src/export/graph.rs

```diff
@@ -0,0 +1,512 @@
+use std::{collections::HashMap, fmt::Write, iter::Peekable};
+
+use serde::Deserialize;
+use uuid::Uuid;
+
+use modality_probe::{EventId, LogicalClock, ProbeId};
+use modality_probe_collector_common::{ReportIter, ReportLogEntry};
+use modality_probe_graph::{EventDigraph, Graph, GraphEvent};
+
+use super::{Cfg, ExportError};
+
+/// A row in the events.csv for a component.
+#[derive(PartialEq, Debug, Clone, Deserialize)]
+pub(super) struct EventMeta {
+    pub component_id: Uuid,
+    pub id: u32,
+    pub name: String,
+    pub type_hint: Option<String>,
+    pub tags: String,
+    pub description: String,
+    pub file: String,
+    pub line: String,
+}
+
+/// A row in probes.csv for a component.
+#[derive(PartialEq, Debug, Clone, Deserialize)]
+pub(super) struct ProbeMeta {
+    pub component_id: Uuid,
+    pub id: u32,
+    pub name: String,
+    pub description: String,
+    pub file: String,
+    pub line: String,
+}
+
+pub(super) fn log_to_graph<I>(
+    log: Peekable<I>,
+) -> Result<EventDigraph<NodeAndEdgeLists<GraphEvent, ()>>, ExportError>
+where
+    I: Iterator<Item = ReportLogEntry>,
+{
+    let mut graph = EventDigraph::new(NodeAndEdgeLists {
+        nodes: HashMap::new(),
+        edges: HashMap::new(),
+    });
+    let report_iter = ReportIter::new(log);
+    for report in report_iter {
+        graph.add_report(&report).map_err(|e| {
+            ExportError(format!(
+                "encountered an error reconstructing the graph: {}",
+                e
+            ))
+        })?;
+    }
+    Ok(graph)
+}
+
+#[derive(Default)]
+pub(super) struct NodeAndEdgeLists<G, W> {
+    nodes: HashMap<G, W>,
+    edges: HashMap<(G, G), W>,
+}
+
+impl NodeAndEdgeLists<GraphEvent, ()> {
+    pub fn into_interactions(self) -> NodeAndEdgeLists<(ProbeId, LogicalClock), u32> {
+        let mut nodes = HashMap::new();
+        for (n, _) in self.nodes.into_iter() {
+            let weight = nodes.entry((n.probe_id, n.clock)).or_insert(0);
+            *weight += 1;
+        }
+
+        let mut edges = HashMap::new();
+        for ((s, t), _) in self.edges.into_iter() {
+            let weight = edges
+                .entry(((s.probe_id, s.clock), (t.probe_id, t.clock)))
+                .or_insert(0);
+            *weight += 1;
+        }
+        NodeAndEdgeLists { nodes, edges }
+    }
+
+    pub fn into_states(self) -> NodeAndEdgeLists<(ProbeId, EventId), u32> {
```

Comment:
Could we have a bit of documentation here on what the states mean here?

Reply:
At a glance, it seems like this is making something like a non-normalized Markov chain representation of event transitions.